Tilers/CityTiler/atlas.py

```diff
@@ -0,0 +1,135 @@
+import sys
+import numpy as np
+from PIL import Image, ImageDraw
+from py3dtiles import BoundingVolumeBox, TriangleSoup
+
+class Rectangle(object):
+
+    def __init__(self, left, top, right, bottom):
+
+        self.left = left
+        self.right = right
+        self.top = top
+        self.bottom = bottom
+        self.width = right - left
+        self.height = bottom - top
+
+    def setSize(self, newWidth, newHeight):
+        self.width = newWidth
+        self.height = newHeight
+
+    def get_top(self):
+        return self.top
+
+    def get_bottom(self):
+        return self.bottom
+
+    def get_right(self):
+        return self.right
+
+    def get_left(self):
+        return self.left
+
+    def get_width(self):
+        return self.width
+
+    def get_height(self):
+        return self.height
+
+    def fits(self,img):
+        width, height = img.size
+        if (width <= (self.width) and (height <= self.height)) :
+            return True
+        else :
+            return False
+
+    def perfect_fits(self,img):
+        width, height = img.size
+        if (width == self.get_width() and height == self.get_height()):
+            return True
+        else : return False
+
+class Node(object):
+
+    def __init__(self,rect = None):
+        self.rect = rect
+        self.child = [None,None]
+        self.image = None
+        self.building_id = None
+
+    def isLeaf(self):
+        return (self.child[0] == None and self.child[1] == None)
+
+    def insertImages(self, atlas , geom):
+        if self.isLeaf():
+
+            if self.image != None :
+                atlas.paste(self.image, (self.rect.get_left(), self.rect.get_top()))
+                self.updateUv(geom[self.building_id].triangles[1], self.image, atlas)
+        else :
+            self.child[0].insertImages(atlas, geom)
+            self.child[1].insertImages(atlas, geom)
+
+
+    def updateUv(self, uvs, oldTexture, newTexture):
+        oldWidth, oldHeight= (oldTexture.size)
+        newWidth, newHeight= (newTexture.size)
+
+        ratioWidth = oldWidth/newWidth
+        ratioHeight = oldHeight/newHeight
+
+        offsetWidth = (self.rect.get_left()/newWidth)
+        offsetHeight = (self.rect.get_top()/newHeight)
+
+        for i in range(0,len(uvs)):
+
+            for y in range(0,3):
+                new_u = ((uvs[i][y][0] * oldWidth) / newWidth) + offsetWidth
+                new_v = ((uvs[i][y][1] * oldHeight) / newHeight) + offsetHeight
+                #warning : don't forget to convert to float32
```

Comment:
Apparently you didn't forget ! Maybe the comment should instead
state why one must convert to float32 ? What was at stake here ?

Reply:
done in c5efbd252ededfa1dfd077572b91b6b3eb4c44d8 , tell me if we need more explanation